runtime/gc_glue_java/ArrayletObjectModel.cpp

```diff
@@ -139,24 +149,70 @@ GC_ArrayletObjectModel::fixupInternalLeafPointersAfterCopy(J9IndexableObject *de
 	if (hasArrayletLeafPointers(destinationPtr)) {
 		GC_ArrayletLeafIterator leafIterator((J9JavaVM*)_omrVM->_language_vm, destinationPtr);
 		GC_SlotObject *leafSlotObject = NULL;
-		UDATA sourceStartAddress = (UDATA) sourcePtr;
-		UDATA sourceEndAddress = sourceStartAddress + getSizeInBytesWithHeader(destinationPtr);
+		uintptr_t sourceStartAddress = (uintptr_t) sourcePtr;
+		uintptr_t sourceEndAddress = sourceStartAddress + getSizeInBytesWithHeader(destinationPtr);
 
 		while (NULL != (leafSlotObject = leafIterator.nextLeafPointer())) {
-			UDATA leafAddress = (UDATA)leafSlotObject->readReferenceFromSlot();
+			uintptr_t leafAddress = (uintptr_t)leafSlotObject->readReferenceFromSlot();
 
 			if ((sourceStartAddress < leafAddress) && (leafAddress < sourceEndAddress)) {
-				leafSlotObject->writeReferenceToSlot((J9Object*)((UDATA)destinationPtr + (leafAddress - sourceStartAddress)));
+				leafSlotObject->writeReferenceToSlot((J9Object*)((uintptr_t)destinationPtr + (leafAddress - sourceStartAddress)));
 			}
 		}
 	}
 }
 
-#if defined(J9VM_ENV_DATA64)
+bool
+GC_ArrayletObjectModel::isArrayletDataAdjacentToHeader(J9IndexableObject *arrayPtr)
+{
+	uintptr_t dataSizeInBytes = getDataSizeInBytes(arrayPtr);
+	return isArrayletDataAdjacentToHeader(dataSizeInBytes);
+}
+
+bool
+GC_ArrayletObjectModel::isArrayletDataAdjacentToHeader(uintptr_t dataSizeInBytes)
+{
+	MM_GCExtensionsBase *extensions = MM_GCExtensionsBase::getExtensions(_omrVM);
+	uintptr_t minimumSpineSizeAfterGrowing = extensions->getObjectAlignmentInBytes();
+	return ((UDATA_MAX == _largestDesirableArraySpineSize) || (dataSizeInBytes <= (_largestDesirableArraySpineSize - minimumSpineSizeAfterGrowing - contiguousIndexableHeaderSize())));
+}
+
+bool
+GC_ArrayletObjectModel::isAddressWithinHeap(MM_GCExtensionsBase *extensions, void *address)
+{
+	uintptr_t heapBase = (uintptr_t)extensions->heap->getHeapBase();
+	uintptr_t heapTop = (uintptr_t)extensions->heap->getHeapTop();
+	return ((uintptr_t)address >= heapBase) && ((uintptr_t)address < heapTop);
+}
+
+bool
+GC_ArrayletObjectModel::isIndexableObjectDoubleMapped(MM_GCExtensionsBase *extensions, J9IndexableObject *arrayPtr)
+{
+#if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+	void *dataAddr = getDataAddrForIndexableObject(arrayPtr);
+	bool isObjectWithinHeap = isAddressWithinHeap(extensions, dataAddr);
+	return ((getDataSizeInBytes(arrayPtr) >= _omrVM->_arrayletLeafSize) && (!isObjectWithinHeap));
```

Comment:
the check in reality is to generic for the very specific name
should we do a more double-map specific check or change the name to be more generic?

Reply:
more generic name is isDataNonadjacentToHeader (picked to somewhat resembled the method above)

the difference between the methods is that the above one consumes size and is used during allocation/construction
while this method is used later (during GC or during  verification) and consumes an already constructed/existing object.

We could name the methods the same name (sDataAdjacentToHeader), but then call sites would have to be adjust to negate the result of the call.